xds/src/main/java/io/grpc/xds/FilterChainMatchingHandler.java

```diff
@@ -0,0 +1,338 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static io.grpc.xds.XdsServerWrapper.ATTR_SERVER_ROUTING_CONFIG;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Strings;
+import com.google.common.collect.Iterables;
+import com.google.protobuf.UInt32Value;
+import io.grpc.Attributes;
+import io.grpc.netty.GrpcHttp2ConnectionHandler;
+import io.grpc.netty.InternalProtocolNegotiationEvent;
+import io.grpc.netty.InternalProtocolNegotiator.ProtocolNegotiator;
+import io.grpc.netty.ProtocolNegotiationEvent;
+import io.grpc.xds.EnvoyServerProtoData.CidrRange;
+import io.grpc.xds.EnvoyServerProtoData.ConnectionSourceType;
+import io.grpc.xds.EnvoyServerProtoData.FilterChain;
+import io.grpc.xds.EnvoyServerProtoData.FilterChainMatch;
+import io.grpc.xds.XdsServerWrapper.ServerRoutingConfig;
+import io.grpc.xds.internal.Matchers.CidrMatcher;
+import io.grpc.xds.internal.sds.SdsProtocolNegotiators.ServerSdsHandler;
+import io.grpc.xds.internal.sds.SslContextProviderSupplier;
+import io.netty.channel.ChannelHandler;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInboundHandlerAdapter;
+import java.net.Inet6Address;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.security.cert.CertStoreException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.Nullable;
+
+
+/**
+ * Handles L4 filter chain match for the connection based on the xds configuration.
+ * */
+public final class FilterChainMatchingHandler extends ChannelInboundHandlerAdapter {
+  private final GrpcHttp2ConnectionHandler grpcHandler;
+  private final FilterChainSelector selector;
+  @Nullable
+  private final ProtocolNegotiator fallbackProtocolNegotiator;
+
+  /**
+   * Selects the filter chain using the selector configuration.
+   * */
+  public FilterChainMatchingHandler(
+          GrpcHttp2ConnectionHandler grpcHandler, FilterChainSelector selector,
+          @Nullable ProtocolNegotiator fallbackProtocolNegotiator) {
+    this.grpcHandler = checkNotNull(grpcHandler, "grpcHandler");
+    this.selector = checkNotNull(selector, "selector");
+    this.fallbackProtocolNegotiator = fallbackProtocolNegotiator;
+  }
+
+  @Override
+  public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exception {
+    if (!(evt instanceof ProtocolNegotiationEvent)) {
+      super.userEventTriggered(ctx, evt);
+      return;
+    }
+    FilteredConfig config = selector.select(
+            (InetSocketAddress) ctx.channel().localAddress(),
+            (InetSocketAddress) ctx.channel().remoteAddress());
+    if (config == null) {
+      throw new IllegalStateException("not filter chain");
```

Comment:
I believe throwing runtime exception is what the current behavior is. Indeed, it looks we should not throw but call `ctx.fireExceptionCaught`? I honestly don't know what would exactly happen if this userEventTriggered throws. But looking at other negotiation handlers ProtocolNegotiationEvent handling, it seems we can just `ctx.fireExceptionCaught(...);return`, and no need to close the `ctx`.

Reply:
If you throw, it is equivalent to calling ctx.fireExceptionCaught() (although calling ctx.fireExceptionCaught() is cheaper than throwing, and a bit more clear). Exceptions propagate and will eventually land in WriteBufferingAndExceptionHandler.executionCaught().

On client-side I guess fireExceptionCaught would yield better results because it can communicate the cause of the connection closure. But we'd want to use a Status\*Exception. On server-side, it probably doesn't matter as there are no queued writes. It seems like nothing will be logged either way, which was my main concern.

I guess this is fine as-is.